chia/wallet/conditions.py

```diff
@@ -0,0 +1,986 @@
+        cls: Type[AssertBeforeHeightAbsolute], program: Program, **kwargs: Any
+    ) -> AssertBeforeHeightAbsolute:
+        return cls(
+            uint32(program.at("rf").as_int()),
+        )
+
+
+@streamable
+@dataclass(frozen=True)
+class Softfork(Condition):
+    cost: uint64
+    conditions: List[Program]
+
+    def to_program(self) -> Program:
+        condition: Program = Program.to([ConditionOpcode.SOFTFORK, self.cost, self.conditions])
+        return condition
+
+    @classmethod
+    def from_program(cls: Type[Softfork], program: Program, **kwargs: Any) -> Softfork:
+        return cls(
+            uint64(program.at("rf").as_int()),
+            list(program.at("rrf").as_iter()),
+        )
+
+
+@streamable
+@dataclass(frozen=True)
+class Remark(Condition):
+    rest: Program
+
+    def to_program(self) -> Program:
+        condition: Program = Program.to((ConditionOpcode.REMARK, self.rest))
+        return condition
+
+    @classmethod
+    def from_program(cls: Type[Remark], program: Program, **kwargs: Any) -> Remark:
+        return cls(
+            program.at("r"),
+        )
+
+
+@streamable
+@dataclass(frozen=True)
+class UnknownCondition(Condition):
+    opcode: Program
+    args: List[Program]
+
+    def to_program(self) -> Program:
+        prog: Program = self.opcode.cons(self.args)
+        return prog
+
+    @classmethod
+    def from_program(cls: Type[UnknownCondition], program: Program, **kwargs: Any) -> UnknownCondition:
+        return cls(
+            program.at("f"), [] if program.at("r") == Program.to(None) else [p for p in program.at("r").as_iter()]
+        )
+
+
+# Abstractions
+@streamable
+@dataclass(frozen=True)
+class AggSig(Condition):
+    pubkey: G1Element
+    msg: bytes
+    opcode: bytes = ConditionOpcode.AGG_SIG_ME.value
+    coin_id: Optional[bytes32] = None
+    parent_id: Optional[bytes32] = None
+    puzzle_hash: Optional[bytes32] = None
+    amount: Optional[uint64] = None
+    additional_data: Optional[bytes32] = None
+
+    def to_program(self) -> Program:
+        # We know we are an agg sig or we want to error
+        return CONDITION_DRIVERS[self.opcode](self.pubkey, self.msg).to_program()  # type: ignore[call-arg]
+
+    @classmethod
+    def from_program(cls: Type[AggSig], program: Program, **kwargs: Any) -> AggSig:
+        opcode: bytes = program.at("f").atom
+        condition_driver: Condition = CONDITION_DRIVERS[opcode].from_program(program, **kwargs)
+        return cls(
+            # We are either parsing an agg sig condition, all of which have these, or we want to error
+            condition_driver.pubkey,  # type: ignore[attr-defined]
+            condition_driver.msg,  # type: ignore[attr-defined]
+            opcode,
+            **{key: value for key, value in condition_driver.__dict__.items() if key not in ["pubkey", "msg"]},
+        )
+
+
+@streamable
+@dataclass(frozen=True)
+class CreateAnnouncement(Condition):
+    msg: bytes
+    coin_not_puzzle: bool
```

Comment:
this could be `source` with `Union[Coin, Puzzle]` ?

Reply:
There's a few reasons that wouldn't work:
1) Streamable can't support unions
2) You don't want to require more information than absolutely necessary.  The only information we need about the origin is a hash (either a coin ID or puzzle hash), and requiring more information could result in us not being able to construct this class when we don't have that information available.
3) The origin info isn't even required anyways, so it could be None and then we would have no way to differentiate.